go/sig3/sig3.go

```diff
@@ -0,0 +1,401 @@
+// one link.
+type RotateKey struct {
+	*Base
+	rkb *RotateKeyBody
+}
+
+var _ Generic = (*RotateKey)(nil)
+
+// NewRotateKey makes a new rotate key given sig3 skeletons (Outer and Inner) and
+// also the PTKs that are going to be advertised in the sig3 link.
+func NewRotateKey(o OuterLink, i InnerLink, b RotateKeyBody) *RotateKey {
+	i.Body = &b
+	return &RotateKey{
+		Base: &Base{
+			inner: &i,
+			outer: o,
+		},
+		rkb: &b,
+	}
+}
+
+func (b *Base) setVerifiedBit() {
+	if b.inner != nil && b.sig != nil {
+		b.verified = true
+	}
+}
+
+// Outer returns a copy of the OuterLink in this base class
+func (b *Base) Outer() OuterLink {
+	return b.outer
+}
+
+func (b *Base) outerPointer() *OuterLink {
+	return &b.outer
+}
+
+// Inner returns a pointer to the InnerLink in this base class.
+func (b *Base) Inner() *InnerLink {
+	return b.inner
+}
+
+// Signer returns the (uid, eldest, KID) of the signer of this link if a sig was provided,
+// and it was successfully verified (as reported by the link itself). If not, then this will
+// return a nil.
+func (b *Base) Signer() *Signer {
+	if !b.verified {
+		return nil
+	}
+	if b.inner == nil {
+		return nil
+	}
+	return &b.inner.Signer
+}
+
+// Prev returns the LinkID of the previous link, or nil if none was provided.
+func (b *Base) Prev() *LinkID {
+	return b.outer.Prev
+}
+
+// Seqno returns the seqno of this link, as reported in the outer link itself.
+func (b *Base) Seqno() Seqno {
+	return b.outer.Seqno
+}
+
+func (b Base) verify() error {
+	if (b.sig == nil) != (b.inner == nil) {
+		return newSig3Error("need sig and inner, or neither, but can't have one and not the other")
+	}
+	if b.sig == nil || b.inner == nil {
+		return nil
+	}
+	err := b.sig.verify(b.inner.Signer.KID, b.outerRaw)
+	if err != nil {
+		return err
+	}
+	b.verified = true
+	return nil
+}
+
+func (i InnerLink) hash() (LinkID, error) {
+	b, err := msgpack.Encode(i)
+	if err != nil {
+		return LinkID{}, err
+	}
+	return hash(b), nil
+}
+
+func (r RotateKey) verify() error {
+	err := r.Base.verify()
+	if err != nil {
+		return err
+	}
+	err = r.verifyReverseSig()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r RotateKey) verifyReverseSig() (err error) {
+	if r.inner == nil {
+		return nil
+	}
+
+	// First make a checkpoint of all of the previous sigs and the previous inner
+	// link, since we're going to mutate them as we verify
+	var reverseSigs []*Sig
+	for _, ptk := range r.rkb.PTKs {
+		reverseSigs = append(reverseSigs, ptk.ReverseSig)
+	}
+	innerLink := r.Base.outer.InnerLink
+
+	// Make sure to replace them on the way out of the function, even in an error.
+	defer func() {
+		for j, rs := range reverseSigs {
+			r.rkb.PTKs[j].ReverseSig = rs
+		}
+		r.Base.outer.InnerLink = innerLink
+	}()
+
+	// Verify signatures in the reverse order they were signed, nulling them out
+	// from back to front. We are not going middle-out.
```

Comment:
:musical_score:

Reply:
This looks fine but is there a reason we don't null out a copy of the link instead so as not to mutate the receiver? Performance?

Reply:
I'd happily use a deep copy operation, but we'd have to write it. I guess we can?

Reply:
I also wonder if the deep copy should keep the aliasing from `RotateKeyBody` into `Base.inner.Body`... Maybe let's punt for now and revisit in the future.